compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SyntheticAccessorLowering.kt

```diff
@@ -443,21 +445,60 @@ internal class SyntheticAccessorLowering(val context: JvmBackendContext) : IrEle
     }
 
     // Counter used to disambiguate accessors to functions with the same base name.
-    private var nameCounter = 0
+    private val nameCounterMap = mutableMapOf<IrFunction, Int>()
+
+    private fun drawNextNameSuffix(function: IrFunction): Int {
+        val next = nameCounterMap[function] ?: 0
+        nameCounterMap[function] = next + 1
+        return next
+    }
 
-    private fun IrFunction.accessorName(): Name {
+    private fun IrFunction.accessorName(superQualifier: IrClassSymbol?): Name {
         val jvmName = context.methodSignatureMapper.mapFunctionName(this)
-        return Name.identifier("access\$$jvmName\$${nameCounter++}")
+        val suffix = when {
+            // The only function accessors placed on interfaces are for JvmDefault implementations
+            parent.safeAs<IrClass>()?.isJvmInterface == true -> "jd"
+
+            // Accessor for _s_uper-qualified call
+            superQualifier != null -> "s" + superQualifier.descriptor.name.asString().hashCode()
+
+            // Access to static members that need an accessor must be because they are inherited,
+            // hence accessed on a _s_upertype.
+            isStatic -> "s" + hashForAccessorDisambiguation()
+
+            else -> drawNextNameSuffix(this)
+        }
+        return Name.identifier("access\$$jvmName\$$suffix")
```

Comment:
Is it possible to avoid `$0` generation? So simple cases would be similar to old backend

Reply:
The test suite says "yes" :) Let me push the revision.